### PR TITLE
DeterministicKeyChain: maybeLookAhead() return unmodifiable, no side-effect

### PR DESCRIPTION
**Pull request by msgilligan**
The private `maybeLookAhead()` methods have been updated to return unmodifiable lists and to have no side effects. The callers previously were responsible for adding new keys to the `basicKeyChain`. Now they are responsible for adding them to both the `basicKeyChain` and the `hierarchy`.

This PR is dependent on PR #2419

### CODE REVIEW

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java [hunk 1]
```diff
@@ -1174,20 +1176,27 @@ public int getLookaheadThreshold() {
...
-            basicKeyChain.importKeys(keys);
+            List<DeterministicKey> keys = concatLists(
+                    maybeLookAhead(externalParentKey, issuedExternalKeys),
+                    maybeLookAhead(internalParentKey, issuedInternalKeys));
+            if (!keys.isEmpty()) {
+                keyLookaheadEpoch++;
+                // Batch add all keys at once so there's only one event listener invocation, as this will be listened to
+                // by the wallet and used to rebuild/broadcast the Bloom filter. That's expensive so we don't want to do
+                // it more often than necessary.
+                basicKeyChain.importKeys(keys);
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:1187]: Are these two lines not also a candidate for wrapping in a method?

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java [hunk 2]
```diff
@@ -505,6 +506,7 @@ public List<DeterministicKey> getKeys(KeyPurpose purpose, int numberOfKeys) {
             // and calculate the full lookahead zone there, so network requests will always use the right amount.
             List<DeterministicKey> lookahead = maybeLookAhead(parentKey, index, 0, 0);
             basicKeyChain.importKeys(lookahead);
+            lookahead.forEach(hierarchy::putKey);
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:509]: Why hasn't this been here before? Is this probably a (hidden) bugfix?

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java (hunk 1 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:1187]: That's the next PR!

Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:1187]: That's the next PR!

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java (hunk 2 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:509]: It was done inside `maybeLookAhead()` -- you can see elsewhere in this PR that I removed it.

Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:509]: It was done inside `maybeLookAhead()` -- you can see elsewhere in this PR that I removed it.

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java (hunk 1 above)
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:1187]: :smile:

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java [hunk 3]
```diff
@@ -1212,10 +1227,7 @@ private List<DeterministicKey> maybeLookAhead(DeterministicKey parent, int issue
         List<DeterministicKey> result = HDKeyDerivation.generate(parent, numChildren)
                 .limit(limit)
                 .map(DeterministicKey::dropPrivateBytes)
-                .collect(Collectors.toList());
-        result.forEach(key -> {
-            hierarchy.putKey(key);
```
Comment (msgilligan) [inline core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:1217]: > Why hasn't this been here before? Is this probably a (hidden) bugfix?

I removed it from here.

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java (hunk 2 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:509]: Note that I also updated the comments on `maybeLookAhead()` to indicate that it is now side effect free.  (This was the major side effect -- adding keys to the hierarchy)

**Comment by schildbach** [comment]
Merged!

**Pull request closed without merging**